### ISSUE TITLE
Naming conventions

### ISSUE THREAD

**Issue by Eastrall** [Eastrall/Sylver#5]
To avoid bad naming conventions, I'd like to open a discussion about them. For example, do we send **messages** or **packets** to the client or server ? Or, on the server-side, do we handle **clients** or **users**.

Let's begin then.

**Note:** Most of the classes of Sylver starts the `Net*` prefix.

## Client or Users ?

The `NetServer` is an abstraction for building a socket server over the TCP/IP protocol. In general, we often ear talking about that a server handles clients. Thing is, Sylver **will** also provide a `NetClient` implementation to builds a socket client that will eventually connect to a `NetServer`.

Now, the question is, a `NetServer` should handle "users" (`NetUser`) or "clients" (`NetClient` -> `NetServerClient` to avoid conflicts?)

## Message or Packet

When I designed the first version of the library (Ether), I based myself on only data structured has packet and has no support for raw data (string or bytes).
Sylver **will** support both `packet` and `raw` data.

Should we rename the `INetPacketStream` interface and implementation to `INetMessageStream` ?

---

Don't hesitate to share your thoughts. 😄

**Comment by Kaev** [comment]
Not all clients are users but all users are clients => Clients
Not all packets are messages but all messages are packets => Packet

**Comment by Eastrall** [comment]
I agree with first point, all users are clients, how do you think we should name the class/interface ? `NetUserClient` ? `NetServerClient`.
Note that a client handled by the server will inherit from this class:
```csharp
public class ChatClient : Net....Client { ... }
public class MyServer : NetServer<ChatClient> { ... }
```

As for the packet and messages, you suggest we keep the packet naming convention?
As for the the `Raw` data, will it be considred as a packet too ?

**Comment by Kaev** [comment]
I would name them just NetClient. If someone wants to extend the class so it's a user or anything besides a NetClient, he can extend it by himself if he wants to.
Yes, i consider raw data as a packet too. I would keep the naming convention.

**Comment by Eastrall** [comment]
I think we will have a problem with namespaces if we use `NetClient` naming for `Sylver.Server` and `Sylver.Client`. It might be confusing to have a 2 `NetClient` classes in the same project and in different namespaces:

- `Sylver.Client`
  - `NetClient`
- `Sylver.Server`
  - `NetServer`
  - `NetClient`

Maybe we should name it : `NetServerClient`. It might be more readable, since we are creating a client for a server instance.

**Comment by carlst99** [comment]
I agree that it would be confusing to have two ```NetClient``` classes. ```NetServerClient``` sounds logical to me, and prevents the issue of using the wrong namespace accidentally when using things like Visual Studio Quick Actions to insert using statements.

Have you considered placing these naming conventions in a contributing file once they are sorted?

**Comment by Eastrall** [comment]
@carlst99 Yes, I am. I don't want to reproduce the same mistakes I did back then with Ether. 😄 

But thanks to your feedbacks about the naming conventions, I will update my code and start adding real markdown documentation about the server part.
Once it is done, I will focus on the `NetClient` side and his documentation.

**Comment by Eastrall** [comment]
Commit 7ed2f9f934c5701d2c95c131bf97e64d2b9637ea adds the `NetServerClient` interface and implementation. From now, we can create a handled client using this code:

```csharp
public class ChatClient : NetServerClient
{
    ...
}

public class MyChatServer : NetServer<ChatClient>
{
    ...
}
```

we can also create a `NetServer` without creating another class:
```csharp
public class ChatClient : NetServerClient { ... }

const string IpAddress = "IP_ADDRESS";
const int Port = 4444;

INetServer myChatServer = new NetServer<ChatClient>(IpAddress, Port);
myChatServer.OnClientConnected += (sender, client) => {
   // do something when a client is connected
};
myChatServer.OnClientDisconnected += (sender, client) => {
   // do something when a client disconnects from the server
};
```
